### PR TITLE
KAFKA-15022: Detect negative cycle from one source

### PR DESCRIPTION
**Pull request by lihaosky**
### Description
Introduce a dummy node connected to every other node and run Bellman-ford from the dummy node once instead of from every node in the graph.

### Testing
Existing test

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java [hunk 1]
```diff
@@ -103,8 +104,24 @@ public String toString() {
         }
     }
 
-    private final SortedMap<V, SortedMap<V, Edge>> adjList = new TreeMap<>();
-    private final SortedSet<V> nodes = new TreeSet<>();
+    private class KeyComparator implements Comparator<V> {
+
+        @Override
+        public int compare(final V o1, final V o2) {
+            if (o1 == null) {
```
Comment (ex172000) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java]: Do we plan to have different results returned under two different null cases? If not, maybe we can do 
```
if (o1 == null || o2 == null) {
    return -1;
}
```
to simplify the condition.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java [hunk 2]
```diff
@@ -200,13 +219,40 @@ public Graph<V> residualGraph() {
...
+            final Edge edge = new Edge(node, 1, 0, 1, 0);
+            destMap.put(node, edge);
+        }
+        residualGraph.adjList.put(null, destMap);
+        residualGraph.nodes.add(null);
+    }
+
+    private void removeDummySourceNode(final Graph<V> residualGraph) {
+        if (!residualGraph.isResidualGraph) {
+            throw new IllegalStateException("Graph should be residual graph to add dummy source node");
```
Comment (ex172000) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java]: Maybe 'Graph should be residual graph to **remove** dummy source node' ?

**Comment by ex172000** [comment]
Will this impact the performance? If so, do we plan to add a performance comparison?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java (hunk 2 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java]: Good catch!

**Comment by lihaosky** [comment]
@ex172000 , this will improve performance. Since it's obvious that we replace the loop of number of nodes to a single iteration, I'm not planning to add comparison.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java [hunk 3]
```diff
@@ -103,8 +104,21 @@ public String toString() {
...
     }
 
-    private final SortedMap<V, SortedMap<V, Edge>> adjList = new TreeMap<>();
-    private final SortedSet<V> nodes = new TreeSet<>();
+    private class KeyComparator implements Comparator<V> {
+
+        @Override
+        public int compare(final V o1, final V o2) {
+            if (o1 == null || o2 == null) {
+                return -1;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java]: If one of both is `null`, why is `o1` considered smaller than `o2`? Would it not be better to define `null` as either smallest or largest globally? -- Also, should we return zero if both are `null`?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java [hunk 4]
```diff
@@ -117,6 +131,8 @@ private Graph(final boolean isResidualGraph) {
     }
 
     public void addEdge(final V u, final V v, final int capacity, final int cost, final int flow) {
+        Objects.requireNonNull(u);
+        Objects.requireNonNull(v);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java:124]: Why do we need these checks?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java (hunk 3 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java]: Good point, if both are `null`, we can return 0;

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java (hunk 4 above)
Comment (lihaosky) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/Graph.java:124]: `null` is a special node we introduced in graph as dummy source to detect negative cycle. So users can use it

**Pull request merged**